README.md
MIMIC-III Benchmarks
=========================

[![Join the chat at https://gitter.im/YerevaNN/mimic3-benchmarks](https://badges.gitter.im/YerevaNN/mimic3-benchmarks.svg)](https://gitter.im/YerevaNN/mimic3-benchmarks?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Python suite to construct benchmark machine learning datasets from the MIMIC-III clinical database. Currently, the benchmark datasets cover four key inpatient clinical prediction tasks that map onto core machine learning problems: prediction of mortality from early admission data (classification), real-time detection of decompensation (time series classification), forecasting length of stay (regression), and phenotype classification (multilabel sequence classification).


## News

* 2018 December 28: The second draft of the paper is released on [arXiv](https://arxiv.org/abs/1703.07771).
* 2017 December 8: This work was presented as a spotlight presentation at NIPS 2017 [Machine Learning for Health Workshop](https://ml4health.github.io/2017/).
* 2017 March 23: We are pleased to announce the first official release of these benchmarks. We expect to release a revision within the coming months that will add at least ~50 additional input variables. We are likewise pleased to announce that the manuscript associated with these benchmarks is now [available on arXiv](https://arxiv.org/abs/1703.07771).

## Citation

If you use this code or these benchmarks in your research, please cite the following publication.
```
@article{Harutyunyan2019,
  author={Harutyunyan, Hrayr and Khachatrian, Hrant and Kale, David C. and Ver Steeg, Greg and Galstyan, Aram},
  title={Multitask learning and benchmarking with clinical time series data},
  journal={Scientific Data},
  year={2019},
  volume={6},
  number={1},
  pages={96},
  issn={2052-4463},
  doi={10.1038/s41597-019-0103-9},
  url={https://doi.org/10.1038/s41597-019-0103-9}
}
```
**Please be sure also to cite the original [MIMIC-III paper](http://www.nature.com/articles/sdata201635).**

## Motivation

Despite rapid growth in research that applies machine learning to clinical data, progress in the field appears far less dramatic than in other applications of machine learning. In image recognition, for example, the winning error rates in the [ImageNet Large Scale Visual Recognition Challenge](http://image-net.org/challenges/LSVRC/) (ILSVRC) plummeted almost 90% from 2010 (0.2819) to 2016 (0.02991).
There are many reasonable explanations for this discrepancy: clinical data sets are [inherently noisy and uncertain](http://www-scf.usc.edu/~dkale/papers/marlin-ihi2012-ehr_clustering.pdf) and often small relative to their complexity, and for many problems of interest, [ground truth labels for training and evaluation are unavailable](https://academic.oup.com/jamia/article-abstract/23/6/1166/2399304/Learning-statistical-models-of-phenotypes-using?redirectedFrom=PDF).

However, there is another, simpler explanation: practical progress has been difficult to measure due to the absence of community benchmarks like ImageNet. Such benchmarks play an important role in accelerating progress in machine learning research. For one, they focus the community on specific problems and stoke ongoing debate about what those problems should be. They also reduce the startup overhead for researchers moving into a new area. Finally and perhaps most important, benchmarks facilitate reproducibility and direct comparison of competing ideas.

Here we present four public benchmarks for machine learning researchers interested in health care, built using data from the publicly available Medical Information Mart for Intensive Care (MIMIC-III) database ([paper](http://www.nature.com/articles/sdata201635), [website](http://mimic.physionet.org)). Our four clinical prediction tasks are critical care variants of four opportunities to transform health care using in "big clinical data" as described in [Bates, et al, 2014](http://content.healthaffairs.org/content/33/7/1123.abstract):

* early triage and risk assessment, i.e., mortality prediction
* prediction of physiologic decompensation
* identification of high-cost patients, i.e. length of stay forecasting
* characterization of complex, multi-system diseases, i.e., acute care phenotyping

In [Harutyunyan, Khachatrian, Kale, and Galstyan 2017](https://arxiv.org/abs/1703.07771), we propose a multitask RNN architecture to solve these four tasks simultaneously and show that this model generally outperforms strong single task baselines.

## Structure
The content of this repository can be divided into four big parts:
* Tools for creating the benchmark datasets.  
* Tools for reading the benchmark datasets.
* Evaluation scripts.
* Baseline models and helper tools.

The `mimic3benchmark/scripts` directory contains scripts for creating the benchmark datasets.
The reading tools are in `mimic3benchmark/readers.py`.
All evaluation scripts are stored in the `mimic3benchmark/evaluation` directory.
The `mimic3models` directory contains the baselines models along with some helper tools.
Those tools include discretizers, normalizers and functions for computing metrics.


## Requirements and installation

We do not provide the MIMIC-III data itself. You must acquire the data yourself from https://mimic.physionet.org/. The required libraries are listed in the `requirements.txt` file.
Here is our suggested way of installation.

1. Install [Miniconda](https://docs.conda.io/en/latest/miniconda.html).
2. Run `conda create -n "mimic3" python=3.7.13`.
3. Run `conda init [SHELL_NAME]`. You might also need to do `source ~/.bashrc` or `source ~/.zshrc` (depends on the shell version).
4. Change the conda environment to `mimic3` by running `conda activate mimic3`.
5. Clone the repo and change the directory:
```bash
       git clone https://github.com/YerevaNN/mimic3-benchmarks/
       cd mimic3-benchmarks/
```
6. Install the requirements with `pip install -r requirements.txt`.

Note that the Python version and some libraries in `requirements.txt` are outdated. This is because the models in `mimic3models/` use an old version of Keras. If you only need to create the benchmark (i.e., you only need the `mimic3benchmark` part), you can use newer Python versions and newer versions of `numpy`, `pandas`, and `scikit-learn`.

## Building the benchmark
Here are the required steps to build the benchmark. It assumes that you already have MIMIC-III dataset (lots of CSV files) on the disk. All the commands below should be run from the root directory of the repo (i.e., from `mimic3-benchmarks`).

1. The following command takes MIMIC-III CSVs, generates one directory per `SUBJECT_ID` and writes ICU stay information to `data/{SUBJECT_ID}/stays.csv`, diagnoses to `data/{SUBJECT_ID}/diagnoses.csv`, and events to `data/{SUBJECT_ID}/events.csv`. This step might take around an hour.
```bash
       python -m mimic3benchmark.scripts.extract_subjects {PATH TO MIMIC-III CSVs} data/root/
```

2. The following command attempts to fix some issues (ICU stay ID is missing) and removes the events that have missing information. About 80% of events remain after removing all suspicious rows (more information can be found in [`mimic3benchmark/scripts/more_on_validating_events.md`](mimic3benchmark/scripts/more_on_validating_events.md)).
```bash
       python -m mimic3benchmark.scripts.validate_events data/root/
```

3. The next command breaks up per-subject data into separate episodes (pertaining to ICU stays). Time series of events are stored in ```{SUBJECT_ID}/episode{#}_timeseries.csv``` (where # counts distinct episodes) while episode-level information (patient age, gender, ethnicity, height, weight) and outcomes (mortality, length of stay, diagnoses) are stores in ```{SUBJECT_ID}/episode{#}.csv```. This script requires two files, one that maps event ITEMIDs to clinical variables and another that defines valid ranges for clinical variables (for detecting outliers, etc.). **Outlier detection is disabled in the current version**.
```bash
       python -m mimic3benchmark.scripts.extract_episodes_from_subjects data/root/
```

4. The next command splits the whole dataset into training and testing sets. Note that the train/test split is the same of all tasks.
```bash
       python -m mimic3benchmark.scripts.split_train_and_test data/root/
```

5. The following commands will generate task-specific datasets, which can later be used in models. These commands are independent, if you are going to work only on one benchmark task, you can run only the corresponding command.
```bash
       python -m mimic3benchmark.scripts.create_in_hospital_mortality data/root/ data/in-hospital-mortality/
       python -m mimic3benchmark.scripts.create_decompensation data/root/ data/decompensation/
       python -m mimic3benchmark.scripts.create_length_of_stay data/root/ data/length-of-stay/
       python -m mimic3benchmark.scripts.create_phenotyping data/root/ data/phenotyping/
       python -m mimic3benchmark.scripts.create_multitask data/root/ data/multitask/
```

After the above commands are done, there will be a directory `data/{task}` for each created benchmark task.
These directories have two subdirectories: `train` and `test`.
Each of them contains a bunch of ICU stays and one file with name `listfile.csv`, which lists all samples in that particular set.
Each row of `listfile.csv` has the following form: `icu_stay, period_length, label(s)`.
A row specifies a sample for which the input is the collection of ICU event of `icu_stay` that occurred in the first `period_length` hours of the stay and the target is/are `label(s)`.
In in-hospital mortality prediction task `period_length` is always 48 hours, so it is not listed in corresponding listfiles.


## Readers
To simplify the reading of benchmark data we wrote special classes.
The `mimic3benchmark/readers.py` contains class `Reader` and five other task-specific classes derived from it.
These are designed to simplify reading of benchmark data. The classes require a directory containing ICU stays and a listfile specifying the samples.
Again, we encourage to use these readers to avoid mistakes in the reading step (for example using events that happened after the first `period_length` hours).  
For more information about using readers view the [`mimic3benchmark/more_on_readers.md`](mimic3benchmark/more_on_readers.md) file.


## Evaluation
For each of the four tasks we provide scripts for evaluating models.
These scripts receive a `csv` file containing the predictions and produce a `json` file containing the scores and confidence intervals for different metrics.
We highly encourage to use these scripts to prevent any mistake in the evaluation step.
For details about the usage of the evaluation scripts view the [`mimic3benchmark/evaluation/README.md`](mimic3benchmark/evaluation/README.md) file.


## Baselines
For each of the four main tasks we provide 7 baselines:  
* Linear/logistic regression
* Standard LSTM
* Standard LSTM + deep supervision
* Channel-wise LSTM
* Channel-wise LSTM + deep supervision
* Multitask standard LSTM
* Multitask channel-wise LSTM

The detailed descriptions of the baselines will appear in the next version of the paper.

Linear models can be found in `mimic3models/{task}/logistic` directories.
LSTM-based models are in `mimic3models/keras_models` directory.

Please note that running linear models can take hours because of extensive grid search and feature extraction.
You can change the size of the training data of linear models in the scripts, so they will become faster (of course the performance will not be the same).

To train the neural networks, you might need to select the keras backend to be `tensorflow`. This can be done by prepending `KERAS_BACKEND=tensorflow` to the training commands below, or by creating an environment variable `KERAS_BACKEND` with value `tensorflow`.

### Train / validation split

Use the following command to extract validation set from the training set. This step is required for running the baseline models. Likewise, the train/test split, the train/validation split is the same for all tasks.
```bash
       python -m mimic3models.split_train_val {dataset-directory}
```
`{dataset-directory}` can be either `data/in-hospital-mortality`, `data/decompensation`, `data/length-of-stay`, `data/phenotyping` or `data/multitask`.


### In-hospital mortality prediction

Run the following command to train the neural network which gives the best result. We got the best performance on validation set after 28 epochs.
```bash
       python -um mimic3models.in_hospital_mortality.main --network mimic3models/keras_models/lstm.py --dim 16 --timestep 1.0 --depth 2 --dropout 0.3 --mode train --batch_size 8 --output_dir mimic3models/in_hospital_mortality
```
Use the following command to train logistic regression. The best model we got used L2 regularization with `C=0.001`:
```bash
       python -um mimic3models.in_hospital_mortality.logistic.main --l2 --C 0.001 --output_dir mimic3models/in_hospital_mortality/logistic
```
### Decompensation prediction

The best model we got for this task was trained for 36 chunks (that's less than one epoch; it overfits before reaching one epoch because there are many training samples for the same patient with different lengths).
```bash
       python -um mimic3models.decompensation.main --network mimic3models/keras_models/lstm.py --dim 128 --timestep 1.0 --depth 1 --mode train --batch_size 8 --output_dir mimic3models/decompensation
```
Use the following command to train a logistic regression. It will have L2 regularization with `C=0.001`, which gave us the best result. To run a grid search over a space of hyper-parameters add `--grid-search` to the command.
```bash
       python -um mimic3models.decompensation.logistic.main --output_dir mimic3models/decompensation/logistic
```
### Length of stay prediction

The best model we got for this task was trained for 19 chunks.
```bash
       python -um mimic3models.length_of_stay.main --network mimic3models/keras_models/lstm.py --dim 64 --timestep 1.0 --depth 1 --dropout 0.3 --mode train --batch_size 8 --partition custom --output_dir mimic3models/length_of_stay
```
Use the following command to train a logistic regression. It will have L1 regularization with `C=0.00001`. To run a grid search over a space of hyperparameters add `--grid-search` to the command.
```bash
       python -um mimic3models.length_of_stay.logistic.main_cf --output_dir mimic3models/length_of_stay/logistic
```
To run a linear regression use this command:
```bash
        python -um mimic3models.length_of_stay.logistic.main --output_dir mimic3models/length_of_stay/logistic
```

### Phenotype classification

The best model we got for this task was trained for 20 epochs.
```bash
       python -um mimic3models.phenotyping.main --network mimic3models/keras_models/lstm.py --dim 256 --timestep 1.0 --depth 1 --dropout 0.3 --mode train --batch_size 8 --output_dir mimic3models/phenotyping
```
Use the following command for logistic regression. It will have L1 regularization with `C=0.1`. To run a grid search over a space of hyperparameters add `--grid-search` to the command.
```bash
       python -um mimic3models.phenotyping.logistic.main --output_dir mimic3models/phenotyping/logistic
```
### Multitask learning

`ihm_C`, `decomp_C`, `los_C` and `ph_C` coefficients control the relative weight of the tasks in the multitask model. Default is `1.0`. Multitask network architectures are stored in `mimic3models/multitask/keras_models`. Here is a sample command for running a multitasking model.
```bash
       python -um mimic3models.multitask.main --network mimic3models/keras_models/multitask_lstm.py --dim 512 --timestep 1 --mode train --batch_size 16 --dropout 0.3 --ihm_C 0.2 --decomp_C 1.0 --los_C 1.5 --pheno_C 1.0 --output_dir mimic3models/multitask
```

## General todos:

- Improve comments and documentation
- Add the best state files for each baseline
- Release 1.0
- Update citation section with Zenodo DOI
- Add to MIMIC's derived data repo
- Refactor, where appropriate, to make code more generally useful
- Expand coverage of variable map and variable range files.
- Decide whether we are missing any other high-priority data (CPT codes, inputs, etc.)


changelog.md
# Changelog


## April 16, 2023

1. Updated tests.


## April 13, 2023


#### Summary
1. Updated pandas and yaml along with code where needed.

2. Using a newer version of scikit-learn.

3. Renamed normalizer names to replace colons ':' with dashes '-'.

4. Fix the requirements.txt file, making sure that both the benchmark and models parts run.

5. Add notes on installation.

#### Renamed normalizer files
The following files are broken in a Windows context because of the `:` in the names. These files were renamed replacing `:` with `-`.
```
./mimic3models/length_of_stay/los_ts0.8.input_str:previous.start_time:zero.n5e4.normalizer
./mimic3models/length_of_stay/los_ts1.0.input_str:previous.start_time:zero.n5e4.normalizer
./mimic3models/in_hospital_mortality/ihm_ts0.8.input_str:previous.start_time:zero.normalizer
./mimic3models/in_hospital_mortality/ihm_ts1.0.input_str:previous.start_time:zero.normalizer
./mimic3models/in_hospital_mortality/ihm_ts2.0.input_str:previous.start_time:zero.normalizer
./mimic3models/multitask/mult_ts1.0.input_str:previous.start_time:zero.normalizer
./mimic3models/phenotyping/ph_ts0.8.input_str:previous.start_time:zero.normalizer
./mimic3models/phenotyping/ph_ts1.0.input_str:previous.start_time:zero.normalizer
./mimic3models/decompensation/decomp_ts0.8.input_str:previous.n1e5.start_time:zero.normalizer
./mimic3models/decompensation/decomp_ts1.0.input_str:previous.n1e5.start_time:zero.normalizer
```

The following commands were issued:
```
 find . -type f -name "*:*" -exec rename 's/:/-/g' {} +
```
Searching for `:.*normalizer` all semicolons in the string were replaced.

models/multitask/main.py
import numpy as np
import argparse
import time
import os
import importlib
import random
import copy

from mimic3benchmark.readers import MultitaskReader
from models.preprocessing import Discretizer, Normalizer
from models import metrics
from models.multitask import utils


parser = argparse.ArgumentParser()
parser.add_argument('--network', type=str, required=True)
parser.add_argument('--dim', type=int, default=256,
                    help='number of hidden units')
parser.add_argument('--epochs', type=int, default=100,
                    help='number of epochs to train')
parser.add_argument('--load_state', type=str, default="",
                    help='state file path')
parser.add_argument('--mode', type=str, default="train",
                    help='mode: train, test or info')
parser.add_argument('--batch_size', type=int, default=64)
parser.add_argument('--l2', type=float, default=0, help='L2 regularization')
parser.add_argument('--l1', type=float, default=0, help='L1 regularization')
parser.add_argument('--log_every', type=int, default=1,
                    help='print information every x iteration')
parser.add_argument('--save_every', type=int, default=1,
                    help='save state every x epoch')
parser.add_argument('--prefix', type=str, default="",
                    help='optional prefix of network name')
parser.add_argument('--no-shuffle', dest='shuffle', action='store_false',
                    help="shuffles the training set")
parser.add_argument('--dropout', type=float, default=0.0, help='dropout rate')
parser.add_argument('--batch_norm', dest='batch_norm', action='store_true',
                    help='batch normalization')
parser.add_argument('--small_part', dest='small_part', action='store_true')
parser.add_argument('--whole_data', dest='small_part', action='store_false')
parser.add_argument('--timestep', type=str, default="1.0",
                    help="fixed timestep used in the dataset")
parser.add_argument('--fm_C', type=float, default=1.0)
parser.add_argument('--los_C', type=float, default=1.0)
parser.add_argument('--ph_C', type=float, default=1.0)
parser.add_argument('--sw_C', type=float, default=1.0)
parser.add_argument('--imputation', type=str, default='previous')
parser.add_argument('--partition', type=str, default="custom", help="log or custom")
parser.add_argument('--nbins', type=int, default=10)

parser.set_defaults(shuffle=True)
parser.set_defaults(batch_norm=True)
parser.set_defaults(small_part=False)
args = parser.parse_args()
print args

train_reader = MultitaskReader(dataset_dir='/home/c9/mimic3-datasets/multitask/train/',
                                  listfile='/home/c9/mimic3-user-datasets/multitask/train_listfile.csv')

val_reader = MultitaskReader(dataset_dir='/home/c9/mimic3-datasets/multitask/train/',
                                 listfile='/home/c9/mimic3-user-datasets/multitask/val_listfile.csv')

discretizer = Discretizer(timestep=float(args.timestep),
                          store_masks=True,
                          imput_strategy='previous',
                          start_time='zero')

discretizer_header = discretizer.transform(train_reader.read_example(0)[0])[1].split(',')
cont_channels = [i for (i, x) in enumerate(discretizer_header) if x.find("->") == -1]

normalizer = Normalizer(fields=cont_channels) # choose here onlycont vs all
normalizer.load_params('mult_ts%s.input_str:%s.start_time:zero.normalizer' % (args.timestep, args.imputation))

train_raw = utils.load_data(train_reader, discretizer, normalizer, args.small_part)
test_raw = utils.load_data(val_reader, discretizer, normalizer, args.small_part)

args_dict = dict(args._get_kwargs())
args_dict['train_raw'] = train_raw
args_dict['test_raw'] = test_raw

# init class
print "==> using network %s" % args.network
network_module = importlib.import_module("networks." + args.network)
network = network_module.Network(**args_dict)
time_step_suffix = ".ts%s" % args.timestep
network_name = args.prefix + network.say_name() + time_step_suffix + "." + args.imputation
print "==> network_name:", network_name


start_epoch = -1
if args.load_state != "":
    start_epoch = network.load_state(args.load_state)

def do_epoch(mode, epoch):
    # mode is 'train' or 'test'

    fm_predictions = []
    fm_answers = []
    
    los_predictions = []
    los_answers = []
    
    ph_predictions = []
    ph_answers = []
    
    sw_predictions = []
    sw_answers = []
    
    avg_loss = 0.0
    sum_loss = 0.0
    prev_time = time.time()
    
    batches_per_epoch = network.get_batches_per_epoch(mode)
    
    for i in range(0, batches_per_epoch):
        step_data = network.step(mode)
        
        fm_pred = step_data["fm_prediction"]
        los_pred = step_data["los_prediction"]
        ph_pred = step_data["ph_prediction"]
        sw_pred = step_data["sw_prediction"]
        
        current_loss = step_data["loss"]
        fm_loss = step_data["fm_loss"]
        los_loss = step_data["los_loss"]
        ph_loss = step_data["ph_loss"]
        sw_loss = step_data["sw_loss"]
        reg_loss = step_data["reg_loss"]
        
        data = step_data["data"]
        
        fm_data = data[1]
        fm_mask = [x[1] for x in fm_data]
        fm_label = [x[2] for x in fm_data]
        
        los_data = data[2]
        los_mask = [x[0] for x in los_data]
        los_label = [x[1] for x in los_data]
        
        ph_data = data[3]
        ph_label = ph_data
        
        sw_data = data[4]
        sw_mask = [x[0] for x in sw_data]
        sw_label = [x[1] for x in sw_data]
        
        avg_loss += current_loss
        sum_loss += current_loss
        
        for (x, mask, y) in zip(fm_pred, fm_mask, fm_label):
            if (mask == 1):
                fm_predictions.append(x)
                fm_answers.append(y)
        
        for (sx, smask, sy) in zip(los_pred, los_mask, los_label):
            for (x, mask, y) in zip(sx, smask, sy):
                if (mask == 1):
                    los_predictions.append(x)
                    los_answers.append(y)

        for (x, y) in zip(ph_pred, ph_label):
            ph_predictions.append(x)
            ph_answers.append(y)
            
        for (sx, smask, sy) in zip(sw_pred, sw_mask, sw_label):
            for (x, mask, y) in zip(sx, smask, sy):
                if (mask == 1):
                    sw_predictions.append(x)
                    sw_answers.append(y)
        
        if ((i + 1) % args.log_every == 0):
            cur_time = time.time()
            print "  {}ing {}.{} / {}  loss: {:8.4f} = {:1.2f} + {:8.2f} + {:1.2f} + "\
                  "{:1.2f} + {:.2f} avg_loss: {:6.4f}  time: {:6.4f}".format(
                        mode, epoch, i * args.batch_size,
                        batches_per_epoch * args.batch_size,
                        float(current_loss),
                        float(fm_loss), float(los_loss), float(ph_loss),
                        float(sw_loss), float(reg_loss),
                        float(avg_loss / args.log_every),
                        float(cur_time - prev_time))
            avg_loss = 0
            prev_time = cur_time
        
        if np.isnan(current_loss):
            print "loss: {:6.4f} = {:1.2f} + {:8.2f} + {:1.2f} + {:1.2f} + {:.2f}".format(
                    float(current_loss),
                    float(fm_loss), float(los_loss), float(ph_loss),
                    float(sw_loss), float(reg_loss))
            raise Exception ("current loss IS NaN. This should never happen :)") 

    sum_loss /= batches_per_epoch
    print "\n  %s loss = %.5f" % (mode, sum_loss)
    
    print "\n ================= 48h mortality ================"
    metrics.print_metrics_binary(fm_answers, fm_predictions)
    
    print "\n ================ length of stay ================"
    if args.partition == 'log':
        metrics.print_metrics_log_bins(los_answers, los_predictions)
    else:
        metrics.print_metrics_custom_bins(los_answers, los_predictions)
        
    print "\n =================== phenotype =================="
    metrics.print_metrics_multilabel(ph_answers, ph_predictions)
    
    print "\n ================ swat mortality ================"
    metrics.print_metrics_binary(sw_answers, sw_predictions)
    
    return sum_loss


if args.mode == 'train':
    print "==> training"   	
    for epoch in range(start_epoch + 1, start_epoch + 1 + args.epochs):
        start_time = time.time()
        
        if args.shuffle:
            network.shuffle_train_set()
        
        do_epoch('train', epoch)
        epoch_loss = do_epoch('test', epoch)

        state_name = 'states/%s.epoch%d.test%.8f.state' % (network_name, epoch,
                                                           epoch_loss)
        if ((epoch + 1) % args.save_every == 0):    
            print "==> saving ... %s" % state_name
            network.save_params(state_name, epoch)
        
        print "epoch %d took %.3fs" % (epoch, float(time.time()) - start_time)

elif args.mode == 'test':
    # ensure that the code uses test_reader
    del train_reader
    del val_reader 
    
    test_reader = MultitaskReader(dataset_dir='/home/c9/mimic3-datasets/multitask/test/',
                                      listfile='/home/c9/mimic3-user-datasets/multitask/test_listfile.csv')
    data_raw = utils.load_data(test_reader, discretizer, normalizer, args.small_part)
    
    data_raw_copy = copy.deepcopy(data_raw) # TODO: delete this 
    
    fm_predictions = []
    fm_answers = []
    
    los_predictions = []
    los_answers = []
    
    ph_predictions = []
    ph_answers = []
    
    sw_predictions = []
    sw_answers = []
    
    avg_loss = 0.0
    sum_loss = 0.0
    prev_time = time.time()
    
    batches_per_epoch = len(data_raw[0]) // args.batch_size
    
    for i in range(0, batches_per_epoch):
        start = i * args.batch_size
        end = start + args.batch_size
        
        data = (data_raw[0][start:end],
                data_raw[1][start:end],
                data_raw[2][start:end],
                data_raw[3][start:end],
                data_raw[4][start:end])
        
        ret = network.predict(data)
        
        data = (data_raw_copy[0][start:end],
                data_raw_copy[1][start:end],
                data_raw_copy[2][start:end],
                data_raw_copy[3][start:end],
                data_raw_copy[4][start:end]) # TODO: delete this
        
        fm_pred = ret["fm_prediction"]
        los_pred = ret["los_prediction"]
        ph_pred = ret["ph_prediction"]
        sw_pred = ret["sw_prediction"]
        
        current_loss = ret["loss"]
        fm_loss = ret["fm_loss"]
        los_loss = ret["los_loss"]
        ph_loss = ret["ph_loss"]
        sw_loss = ret["sw_loss"]
        reg_loss = ret["reg_loss"]
        
        fm_data = data[1]
        fm_mask = [x[1] for x in fm_data]
        fm_label = [x[2] for x in fm_data]
        
        los_data = data[2]
        los_mask = [x[0] for x in los_data]
        los_label = [x[1] for x in los_data]
        
        ph_data = data[3]
        ph_label = ph_data
        
        sw_data = data[4]
        sw_mask = [x[0] for x in sw_data]
        sw_label = [x[1] for x in sw_data]
        
        avg_loss += current_loss
        sum_loss += current_loss
        
        for (x, mask, y) in zip(fm_pred, fm_mask, fm_label):
            if (mask == 1):
                fm_predictions.append(x)
                fm_answers.append(y)
        
        for (sx, smask, sy) in zip(los_pred, los_mask, los_label):
            for (x, mask, y) in zip(sx, smask, sy):
                if (mask == 1):
                    los_predictions.append(x)
                    los_answers.append(y)

        for (x, y) in zip(ph_pred, ph_label):
            ph_predictions.append(x)
            ph_answers.append(y)
            
        for (sx, smask, sy) in zip(sw_pred, sw_mask, sw_label):
            for (x, mask, y) in zip(sx, smask, sy):
                if (mask == 1):
                    sw_predictions.append(x)
                    sw_answers.append(y)
        
        if ((i + 1) % args.log_every == 0):
            cur_time = time.time()
            print "  {}ing {} / {}  loss: {:8.4f} = {:1.2f} + {:8.2f} + {:1.2f} + "\
                  "{:1.2f} + {:.2f} avg_loss: {:6.4f}  time: {:6.4f}".format(
                        args.mode, i * args.batch_size,
                        batches_per_epoch * args.batch_size,
                        float(current_loss),
                        float(fm_loss), float(los_loss), float(ph_loss),
                        float(sw_loss), float(reg_loss),
                        float(avg_loss / args.log_every),
                        float(cur_time - prev_time))
            avg_loss = 0
            prev_time = cur_time
        
        if np.isnan(current_loss):
            print "loss: {:6.4f} = {:1.2f} + {:8.2f} + {:1.2f} + {:1.2f} + {:.2f}".format(
                    float(current_loss),
                    float(fm_loss), float(los_loss), float(ph_loss),
                    float(sw_loss), float(reg_loss))
            raise Exception ("current loss IS NaN. This should never happen :)") 

    sum_loss /= batches_per_epoch
    print "\n  %s loss = %.5f" % (args.mode, sum_loss)
    
    print "\n ================= 48h mortality ================"
    metrics.print_metrics_binary(fm_answers, fm_predictions)
    
    print "\n ================ length of stay ================"
    if args.partition == 'log':
        metrics.print_metrics_log_bins(los_answers, los_predictions)
    else:
        metrics.print_metrics_custom_bins(los_answers, los_predictions)
    
    print "\n =================== phenotype =================="
    metrics.print_metrics_multilabel(ph_answers, ph_predictions)
    
    print "\n ================ swat mortality ================"
    metrics.print_metrics_binary(sw_answers, sw_predictions)
    
    with open("los_activations.txt", "w") as fout:
        fout.write("prediction, y_true")
        for (x, y) in zip(los_predictions, los_answers):
            fout.write("%.6f, %.6f\n" % (x, y))
    
else:
    raise Exception("unknown mode")